frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2099,6 +2119,11 @@ const ResolvedFunction* resolveConcreteFunction(Context* context, ID id) {
 
   const UntypedFnSignature* uSig = UntypedFnSignature::get(context, id);
   const TypedFnSignature* sig = typedSignatureInitial(context, uSig);
+
+  if (sig == nullptr) {
```

Comment:
Two things:
1. This and other nullptr checks for `typedSignatureInitial` results; is their inclusion because of new `nullptr`s coming from a default `nullptr` when handling a recursive query? Or should they have been present before too
2. Small thing: all the other `nullptr` checks were included into the `needsInstantiation` conditionals. Perhaps do the same with the conditional below? `if (sig == nullptr || sig->needsInstantiation())`

Reply:
>  is their inclusion because of new nullptrs coming from a default nullptr when handling a recursive query? Or should they have been present before too?

It's because of the default `nullptr` during recursion. It wasn't possible for `typedSignatureInitial` to return `nullptr` before, since it's very light on "argument-passing" and other logic. However, if resolution of formal types hits recursion, `nullptr` is now on the table.

 > Perhaps do the same with the conditional below? if (sig == nullptr || sig->needsInstantiation())

Sure, why not? 😄